Dockerfile

```diff
@@ -1,15 +1,32 @@
 # Choose your LLVM version
-ARG LLVM_VERSION=8.0
+ARG LLVM_VERSION=800
+ARG ARCH=amd64
+ARG DISTRO_BASE=ubuntu18.04
+ARG LIBRARIES=/opt/trailofbits/libraries
+ARG Z3_VERSION=4.7.1
+#this seems to be static for most Z3 Linux builds
+ARG Z3_OS_VERSION=ubuntu-16.04
+ARG Z3_ARCHIVE=z3-${Z3_VERSION}-x64-${Z3_OS_VERSION}
 
 # Run-time dependencies go here
-FROM ubuntu:18.04 as base
+FROM trailofbits/cxx-common:llvm${LLVM_VERSION}-${DISTRO_BASE}-${ARCH} as base
+ARG Z3_ARCHIVE
+ARG Z3_VERSION
+ARG LIBRARIES
+#FROM ubuntu:18.04 as base
 RUN apt-get update && \
     apt-get upgrade -y && \
     apt-get install -y --no-install-recommends \
-     libgomp1 && \
+     libgomp1 unzip curl && \
     apt-get clean && \
     rm -rf /var/lib/apt/lists/*
 
+#TODO: replace with git clone & build of the version to work on non-amd64 arch
+RUN curl -OL "https://github.com/Z3Prover/z3/releases/download/z3-${Z3_VERSION}/${Z3_ARCHIVE}.zip" && \
+    unzip -qq "${Z3_ARCHIVE}.zip" && \
+    mv "${Z3_ARCHIVE}" "${LIBRARIES}/z3" && \
+    rm "${Z3_ARCHIVE}.zip"
+
 # Build-time dependencies go here
 FROM base as deps
```

Comment:
We could think about putting `cxx-common` here, as part of the build dependencies, which won't be included in the final image.